# 45장: 프로미스

> "ES6에서... 프로미스를 도입했다. 프로미스는 전통적인 콜백 패턴이 가진 단점을 보완하며 비동기 처리 시점을 명확하게 표현할 수 있다는 장점이 있다" (p.842)

-   [45장: 프로미스](#45장-프로미스)
    -   [45-1. 비동기 처리를 위한 콜백 패턴의 단점](#45-1-비동기-처리를-위한-콜백-패턴의-단점)
        -   [45-1-1. 콜백 헬Callback Hell](#45-1-1-콜백-헬callback-hell)
        -   [45-1-2. 에러 처리의 한계](#45-1-2-에러-처리의-한계)
    -   [45-2. 프로미스의 생성](#45-2-프로미스의-생성)
        -   [프로미스 객체 내부](#프로미스-객체-내부)
    -   [45-3. 프로미스의 후속 처리 메서드](#45-3-프로미스의-후속-처리-메서드)
        -   [Promise.prototype.then](#promiseprototypethen)
        -   [Promise.prototype.catch](#promiseprototypecatch)
        -   [Promise.prototype.finally](#promiseprototypefinally)
    -   [45-4. 프로미스의 에러 처리](#45-4-프로미스의-에러-처리)
    -   [45-5. 프로미스 체이닝](#45-5-프로미스-체이닝)
    -   [45-7. 마이크로태스크 큐](#45-7-마이크로태스크-큐)

---

## 45-1. 비동기 처리를 위한 콜백 패턴의 단점

### 45-1-1. 콜백 헬<sup>Callback Hell</sup>

```javascript
get("/step1", (a) => {
    get("/step2", (b) => {
        get("/step3", (c) => {
            get("/step4", (d) => {
                console.log(d);
            });
        });
    });
});
```

-   비동기 함수는 비동기 처리 결과를 외부에 반환할 수 업소, 상위 스코프의 변수에 할당할 수도 없다
-   따라서 비동기 함수의 처리 결과(서버의 응답 등)에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다
-   이때 비동기 함수를 범용적으로 사용하기 위해 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백 함수를 전달하는 것이 일반적이다
-   하지만 그렇기 때문에 비동기 함수가 비동기 처리 결과를 가지고 또 다시 비동기 함수를 호출해야 하는 경우가 있으면 **콜백 헬**이 된다
-   콜백 헬은 가독성이 나쁠뿐 아니라 실수/에러가 발생할 확률이 높다

### 45-1-2. 에러 처리의 한계

```javascript
try {
    setTimeout(() => {
        throw new Error("Intentional Error.");
    }, 1000);
} catch (error) {
    console.error("Caught this error: ", e);
}
```

-   '비동기 처리를 위한 콜백 패턴'의 문제점 중에서 가장 심각한 것은 에러 처리가 곤란하다는 점이다
-   에러는 호출자 방향으로 전파된다. 즉, 콜스택의 아래 방향(실행 중인 실행 컨텍스트가 push되기 직전에 push된 실행 컨텍스트 방향)으로 전파된다
-   위 예제에서 setTimeout 함수를 통해 1초 후 고의적인 에러를 발생시켰다
-   하지만 setTimeout 함수의 콜백 함수를 호출한 것은 setTimeout 함수가 아니다, 따라서 setTimeout 함수의 콜백 함수가 발생시킨 에러는 catch 블록에서 캐치되지 않는다

<br>

## 45-2. 프로미스의 생성

```javascript
const promise = new Promise((resolve, reject) => {
  if(/* 비동기 처리 성공 */) resolve('result');
  else rejejct('failure reason');
});
```

-   Promise 생성자 함수를 new 연산자와 함께 호출하면 프로미스 객체를 생성한다 (표준 빌트인)
-   Promise 생성자 함수는 비동기 처리를 수행할 콜백 함수<sup>Executor</sup>를 인수로 받는데, 이 콜백 함수는 resolve와 reject 함수를 인수로 전달받는다
-   비동기 처리가 성공적이면 'resolve'함수, 비동기 처리가 실패하면 'reject' 함수를 호출한다
-   프로미스 객체는 다음과 같이 현재 비동기 처리가 어떻게 진행되고 있는지를 나타내는 상태 정보를 갖는다

| <center>프로미스의 상태 정보</center> | <center>의미</center>                 | <center>상태 변경 조건</center>  |
| ------------------------------------- | ------------------------------------- | -------------------------------- |
| pending                               | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfilled                             | 비동기 처리가 수행된 상태(성공)       | resolve 함수 호출                |
| rejected                              | 비동기 처리가 수행된 상태(실패)       | reject 함수 호출                 |

-   성공<sup>fulfilled</sup>, 실패<sup>rejected</sup> 상태를 합쳐 **settled**라고 표현한다
-   pending 상태에서는 settled 상태로 변화할 수 있지만, 한번 settled 상태가 되면 더 이상 다른 상태로 변화할 수 없다

### 프로미스 객체 내부

```javascript
const fulfilledPromise = new Promise((resolve) => resolve(1));
```

-   위 코드를 개발자 도구에서 돌려보면 `[[PromiseStatus]]`에 fulfilled, `[[PromiseValue]]`에 1이 저장되어 있는 것을 확인할 수 있다 (reject의 경우 마찬가지지만 `[[PromiseValue]]`에 Error 정보 저장되어 있음)
-   즉, 프로미스 개체는 비동기 처리 상태와 처리 결과를 관리하는 객체라 볼 수 있다

<br>

## 45-3. 프로미스의 후속 처리 메서드

```javascript
const examplePromise = new Promise((resolve) => resolve("fulfiled"))
    .then((firstLineResult) => console.log(firstLineResult))
    .catch((error) => console.log(error))
    .finally(() => console.log("Bye!"));

/* Output =>
fulfilled
Bye!
*/
```

-   프로미스의 비동기 처리 상태가 변화하면 이에 따른 후속 처리를 해야 한다
    -   "fulfilled"의 경우 결과를 가지고 무언가를 한다
    -   "rejected"의 경우 에러 처리를 해야 한다
-   프로미스는 이를 위한 '후속 처리 메서드' `then`, `catch`, `finally`를 제공한다

### Promise.prototype.then

-   `then` 메서드는 두 개의 콜백 함수를 인수로 받는다. 첫째 콜백 함수는 비동기 처리가 성공<sup>fulfilled</sup>한 경우, 둘째 콜백 함수는 처리가 실패<sup>rejected</sup>된 경우 호출된다
-   통상 fulfilled인 경우의 처리만 담당하게끔 사용한다

### Promise.prototype.catch

-   `catch` 메서드는 하나의 콜백 함수를 인수로 받고, 이는 비동기 처리가 실패한 경우에만 호출된다
-   `then(_, onRejectedCallbackFunc)`와 동일하게 동작한다고 보면 된다

### Promise.prototype.finally

-   `finally` 메서드는 하나의 콜백 함수를 인수로 받고, 이는 fulfilled/rejected 상관없이 무조건 한 번 호출된다

<br>

## 45-4. 프로미스의 에러 처리

-   '비동기 처리를 위한 콜백 패턴'은 에러 처리가 곤란하다는 단점이 있었다
-   하지만 Promise를 사용하면 then, catch, finally 메서드를 활용해서 에러를 문제 없이 처리할 수 있다

```javascript
const wrongUrl = "https://www.thisisnotavalidurl.com";

promiseGet(wrongUrl).then(
    (res) => console.xxx(res),
    (err) => console.log(err)
);
```

-   주의해야할 점이 있다면 then 메서드의 두 번째 인수로 rejected 상태일 시 실행할 콜백 함수를 전달할 수 있다. 그러나 이렇게 사용하는 경우 콜백 함수 내부에서 발생하는 에러를 잡을 수 없다
-   위 예제처럼 에러 처리를 then의 두 번째 인수로 처리하는 경우 `console.xxx` 같은 말도 안되는 코드를 작성했음에도 에러가 캐치되지 않으며, 가독성도 나빠진다
-   따라서 then 메서드로는 fulfilled 처리만 하고, 에러 캐치는 catch로 하는 것이 가독성도 좋고 명확하다

<br>

## 45-5. 프로미스 체이닝

-   '비동기 처리를 위한 콜백 패턴'은 연속된 비동기 처리를 하기 위해서 **콜백 헬**에 빠지는 단점이 있었다

```
💡 promise --> then --> then --> then --> then --> catch --> finally
```

-   Promise의 후속처리 메서드<sup>then, catch, finally</sup>는 모두 프로미스를 반환하기 때문에 **프로미스 체이닝**이 가능하다. 즉, then 메서드를 연속적으로 호출할 수 있다
-   단, 프로미스도 콜백 패턴을 사용하므로 콜백 함수의 사용이 아예 사라진 건 아니다. 이는 가독성이 좋지 않으며 추후 배울 `async/await`를 사용해서 해결할 수 있다

<br>

## 45-7. 마이크로태스크 큐

```javascript
setTimeout(() => console.log("Timeout Callback Func"), 0);

Promise.resolve()
    .then(() => console.log("first then callback func"))
    .then(() => console.log("second then callback func"));

/* Output -> 
first then callback func
second then callback func 
Timeout Callback Func
*/
```

-   위 예제에서 Promise도 비동기 처리이기 때문에 Timeout Callback Func부터 순서대로 출력될 것 같지만 실제로는 Promise의 then 메서드 처리부터 되고 setTimeout의 콜백 함수가 실행된다
-   이는 Promise의 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 마이크로태스크 큐<sup>Microtask Queue/Job Queue</sup>에 저장되기 때문이다
-   마이크로태스크 큐는 태스크 큐보다 우선순위가 높은 별도의 큐다. 즉, 콜 스택이 비면 먼저 마이크로태스크 큐에서 대기하고 있는 함수를 가져와 실행하고, 그 다음 태스크 큐에서 대기하고 있는 함수를 가져온다
